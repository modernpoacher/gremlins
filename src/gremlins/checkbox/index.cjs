require('@babel/register')({
  ignore: [
    /node_modules\/(?!@modernpoacher\/cogs|@modernpoacher\/sprockets|@modernpoacher\/gremlins)/
  ]
})

const debug = require('debug')

const log = debug('@modernpoacher/gremlins/gremlins/textbox')

log('`gremlins` is awake')

const {
  default: component
} = require('./index.jsx')

module.exports = component
