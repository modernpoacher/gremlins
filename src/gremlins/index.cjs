require('@babel/register')({
  ignore: [
    /node_modules\/(?!@modernpoacher\/cogs|@modernpoacher\/gremlins|@modernpoacher\/sprockets)/
  ]
})

const debug = require('debug')

const log = debug('@modernpoacher/gremlins/gremlins')

log('`gremlins` is awake')

/**
 *  Has default export, etc
 */
module.exports = require('./index.jsx')
