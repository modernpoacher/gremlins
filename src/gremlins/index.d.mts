declare module '#gremlins/gremlins' {
  import React from 'react'

  export interface FieldGremlinProps {
    id: string
    name: string
    required: boolean
    disabled: boolean
    readOnly: boolean
    placeholder: string
    onChange: GremlinsTypes.OnChangeType
    fieldRef: object
  }

  export interface ValueGremlinProps extends FieldGremlinProps {
    value: string
    defaultValue: string
    onChange: GremlinsTypes.OnChangeType
  }

  export interface CheckGremlinProps extends FieldGremlinProps {
    checked: boolean
    defaultChecked: boolean
    onClick: GremlinsTypes.OnClickType
  }

  export interface GroupGremlinProps {
    title: string
    onChange: GremlinsTypes.OnChangeType
    groupRef: object
    children: React.JSX.Element | React.JSX.Element[]
  }

  export class FieldGremlin extends React.Component<FieldGremlinProps> {}

  export class ValueGremlin extends React.Component<ValueGremlinProps> {}

  export class CheckGremlin extends React.Component<CheckGremlinProps> {}

  export class GroupGremlin extends React.Component<GroupGremlinProps> {}
}

declare module '@modernpoacher/gremlins/gremlins' {
  export * from '#gremlins/gremlins'
}
