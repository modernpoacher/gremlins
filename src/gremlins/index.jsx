/**
 * GroupGremlin component
 *
 * @typedef {import('@modernpoacher/gremlins/gremlins').GroupGremlinProps} GroupGremlinProps
 */
/**
 * FieldGremlin component
 *
 * @typedef {import('@modernpoacher/gremlins/gremlins').FieldGremlinProps} FieldGremlinProps
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import {
  DEFAULT_HANDLE_CHANGE,
  DEFAULT_HANDLE_CLICK
} from '@modernpoacher/gremlins/common'

import Field, {
  ValueField,
  CheckField
} from '@modernpoacher/gremlins/components/field'

import Group from '@modernpoacher/gremlins/components/group'

export class FieldGremlin extends Component {
  getClassName () {
    const {
      required = false,
      disabled = false,
      readOnly = false
    } = this.props

    return classnames('gremlin', { required, disabled, readOnly })
  }

  getId () {
    const {
      id,
      name
    } = this.props

    return id || name
  }

  /**
   * @param {FieldGremlinProps} props
   * @returns {boolean}
   */
  shouldComponentUpdate (props) {
    return (
      (props.name !== this.props.name) ||
      (props.id !== this.props.id) ||
      (props.required !== this.props.required) ||
      (props.disabled !== this.props.disabled) ||
      (props.readOnly !== this.props.readOnly) ||
      (props.tabIndex !== this.props.tabIndex) ||
      (props.accessKey !== this.props.accessKey) ||
      (props.placeholder !== this.props.placeholder) ||
      (props.onChange !== this.props.onChange)
    )
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      required = false,
      disabled = false,
      readOnly = false,
      tabIndex,
      accessKey,
      placeholder,
      onChange = DEFAULT_HANDLE_CHANGE,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={onChange}
        fieldRef={fieldRef}
      />
    )
  }

  render () {
    const className = this.getClassName()

    return (
      <div className={className}>
        {this.renderField()}
      </div>
    )
  }
}

FieldGremlin.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
  fieldRef: PropTypes.shape({
    current: PropTypes.shape().isRequired
  })
}

FieldGremlin.defaultProps = {
  required: false,
  disabled: false,
  readOnly: false,
  onChange: DEFAULT_HANDLE_CHANGE
}

export class ValueGremlin extends FieldGremlin {
  shouldComponentUpdate (props) {
    return (
      super.shouldComponentUpdate(props) ||
      (props.value !== this.props.value)
    )
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      required = false,
      disabled = false,
      readOnly = false,
      tabIndex,
      accessKey,
      placeholder,
      onChange = DEFAULT_HANDLE_CHANGE,
      fieldRef
    } = this.props

    return (
      <ValueField
        id={id}
        name={name}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={onChange}
        fieldRef={fieldRef}
      />
    )
  }
}

ValueGremlin.propTypes = {
  ...FieldGremlin.propTypes,
  value: PropTypes.string,
  defaultValue: PropTypes.string
}

ValueGremlin.defaultProps = {
  ...FieldGremlin.defaultProps
}

export class CheckGremlin extends FieldGremlin {
  shouldComponentUpdate (props) {
    return (
      super.shouldComponentUpdate(props) ||
      (props.checked !== this.props.checked) ||
      (props.onClick !== this.props.onClick)
    )
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      required = false,
      disabled = false,
      readOnly = false,
      tabIndex,
      accessKey,
      placeholder,
      onChange = DEFAULT_HANDLE_CHANGE,
      onClick = DEFAULT_HANDLE_CLICK,
      fieldRef
    } = this.props

    return (
      <CheckField
        id={id}
        name={name}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={onChange}
        onClick={onClick}
        fieldRef={fieldRef}
      />
    )
  }
}

CheckGremlin.propTypes = {
  ...FieldGremlin.propTypes,
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  onClick: PropTypes.func
}

CheckGremlin.defaultProps = {
  ...FieldGremlin.defaultProps,
  onClick: DEFAULT_HANDLE_CLICK
}

export class GroupGremlin extends Component {
  getClassName () {
    return 'gremlin'
  }

  /**
   * @param {GroupGremlinProps} props
   * @returns {boolean}
   */
  shouldComponentUpdate (props) {
    return (
      (props.children !== this.props.children) ||
      (props.onChange !== this.props.onChange)
    )
  }

  renderGroup () {
    const {
      onChange = DEFAULT_HANDLE_CHANGE,
      groupRef,
      children
    } = this.props

    return (
      <Group
        onChange={onChange}
        groupRef={groupRef}>
        {children}
      </Group>
    )
  }

  render () {
    const className = this.getClassName()

    return (
      <div className={className}>
        {this.renderGroup()}
      </div>
    )
  }
}

GroupGremlin.propTypes = {
  onChange: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(
      PropTypes.node
    )
  ]),
  groupRef: PropTypes.shape({
    current: PropTypes.shape().isRequired
  })
}

GroupGremlin.defaultProps = {
  onChange: DEFAULT_HANDLE_CHANGE
}
