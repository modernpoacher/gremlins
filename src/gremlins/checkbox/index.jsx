/**
 * CheckboxGremlin component
 */
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { CheckGremlin } from '#gremlins/gremlins'

import {
  DEFAULT_HANDLE_CHANGE,
  DEFAULT_HANDLE_CLICK
} from '#gremlins/common'

import Field from './field/index.jsx'

export default class CheckboxGremlin extends CheckGremlin {
  getClassName () {
    return classnames(super.getClassName(), 'checkbox')
  }

  handleChange = (value, checked) => {
    const {
      onChange = DEFAULT_HANDLE_CHANGE
    } = this.props

    onChange(value, checked)
  }

  handleClick = (value, checked) => {
    const {
      onClick = DEFAULT_HANDLE_CLICK
    } = this.props

    onClick(value, checked)
  }

  shouldComponentUpdate (props, state) {
    return (
      super.shouldComponentUpdate(props) ||
      (props.value !== this.props.value)
    )
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      value,
      checked,
      defaultChecked,
      required = false,
      disabled = false,
      readOnly = false,
      tabIndex,
      accessKey,
      placeholder,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        value={value}
        checked={checked}
        defaultChecked={defaultChecked}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={this.handleChange}
        onClick={this.handleClick}
        fieldRef={fieldRef}
      />
    )
  }

  render () {
    const className = this.getClassName()

    return (
      <div className={className}>
        {this.renderField()}
      </div>
    )
  }
}

CheckboxGremlin.propTypes = {
  ...CheckGremlin.propTypes,
  value: PropTypes.string.isRequired
}
