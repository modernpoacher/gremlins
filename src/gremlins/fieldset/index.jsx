/**
 * FieldsetGremlin component
 *
 * @typedef {import('@modernpoacher/gremlins/gremlins').GroupGremlinProps} GroupGremlinProps
 */

import React from 'react'

import classnames from 'classnames'

import { GroupGremlin as Gremlin } from '#gremlins/gremlins'

import {
  DEFAULT_HANDLE_CHANGE
} from '#gremlins/common'

import Group from './group/index.jsx'

export default class FieldsetGremlin extends Gremlin {
  getClassName () {
    return classnames(super.getClassName(), 'fieldset')
  }

  renderGroup () {
    const {
      onChange = DEFAULT_HANDLE_CHANGE,
      children
    } = this.props

    return (
      <Group
        onChange={onChange}>
        {children}
      </Group>
    )
  }
}

FieldsetGremlin.propTypes = {
  ...Gremlin.propTypes
}
