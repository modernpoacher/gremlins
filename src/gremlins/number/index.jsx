/**
 * NumberGremlin component
 */
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { ValueGremlin } from '#gremlins/gremlins'

import {
  DEFAULT_HANDLE_CHANGE
} from '#gremlins/common'

import Field from './field/index.jsx'

export default class NumberGremlin extends ValueGremlin {
  getClassName () {
    return classnames(super.getClassName(), 'number')
  }

  handleChange = (value) => {
    const {
      onChange = DEFAULT_HANDLE_CHANGE
    } = this.props

    onChange(value)
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      value,
      defaultValue,
      required = false,
      disabled = false,
      readOnly = false,
      tabIndex,
      accessKey,
      placeholder,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={this.handleChange}
        fieldRef={fieldRef}
      />
    )
  }
}

NumberGremlin.propTypes = {
  ...ValueGremlin.propTypes,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  defaultValue: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ])
}
