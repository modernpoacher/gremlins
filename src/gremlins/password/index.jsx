/**
 * PasswordGremlin component
 */
import React from 'react'
import classnames from 'classnames'

import { ValueGremlin } from '@modernpoacher/gremlins/gremlins'

import Field from './field/index.jsx'

export default class PasswordGremlin extends ValueGremlin {
  getClassName () {
    return classnames(super.getClassName(), 'password')
  }

  handleChange = (value) => {
    const {
      onChange
    } = this.props

    onChange(value)
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      value,
      defaultValue,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      placeholder,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={this.handleChange}
        fieldRef={fieldRef}
      />
    )
  }
}

PasswordGremlin.propTypes = {
  ...ValueGremlin.propTypes
}

PasswordGremlin.defaultProps = {
  ...ValueGremlin.defaultProps
}
