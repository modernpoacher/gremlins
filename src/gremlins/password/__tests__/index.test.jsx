import React, { Component as mockComponent } from 'react'
import renderer from 'react-test-renderer'
import classnames from 'classnames'

import { ValueGremlin } from '@modernpoacher/gremlins/gremlins'
import Gremlin from '@modernpoacher/gremlins/gremlins/password'

import Field from '../field/index.jsx'

jest.mock('classnames', () => jest.fn(() => 'MOCK CLASSNAME'))

jest.mock('@modernpoacher/gremlins/gremlins', () => {
  class MockGremlin extends mockComponent {
    getClassName () { }

    getId () {}

    renderField () {
    }

    render () {
      const className = this.getClassName()

      return (
        <div className={className}>
          {this.renderField()}
        </div>
      )
    }
  }

  return {
    __esModule: true,
    ValueGremlin: class ValueGremlin extends MockGremlin { },
    default: MockGremlin
  }
})

jest.mock('../field/index.jsx')

describe('@modernpoacher/gremlins/gremlins/password', () => {
  describe('<Gremlin />', () => {
    describe('With required props', () => {
      const component = (
        <Gremlin name='MOCK NAME' />
      )

      it('renders', () => {
        return expect(renderer.create(component).toJSON())
          .toMatchSnapshot()
      })

      describe('`getClassName`', () => {
        it('is defined', () => {
          return expect(Gremlin.prototype.getClassName)
            .toBeDefined()
        })
      })

      describe('`renderField`', () => {
        it('is defined', () => {
          return expect(Gremlin.prototype.renderField)
            .toBeDefined()
        })
      })
    })

    describe('With additional props', () => {
      it('renders', () => {
        const component = (
          <Gremlin
            name='MOCK NAME'
            id='MOCK ID'
            value='MOCK VALUE'
            tabIndex={1}
            accessKey='MOCK ACCESS KEY'
            required
            disabled
            readOnly
            placeholder='MOCK PLACEHOLDER'
            onChange={jest.fn()}
          />
        )

        return expect(renderer.create(component).toJSON())
          .toMatchSnapshot()
      })
    })

    describe('`getClassName()`', () => {
      let returnValue

      beforeEach(() => {
        jest.spyOn(ValueGremlin.prototype, 'getClassName').mockReturnValue('MOCK GETCLASSNAME')

        const component = (
          <Gremlin name='MOCK NAME' />
        )

        const instance = (
          renderer.create(component)
            .getInstance()
        )

        returnValue = instance.getClassName()
      })

      it('does not invoke `classnames`', () => {
        return expect(classnames)
          .toBeCalledWith('MOCK GETCLASSNAME', 'password')
      })

      it('returns the classname', () => {
        return expect(returnValue)
          .toBe('MOCK CLASSNAME')
      })
    })

    describe('`renderField()`', () => {
      const component = (
        <Gremlin
          name='MOCK NAME'
          id='MOCK ID'
          value='MOCK VALUE'
          tabIndex={1}
          accessKey='MOCK ACCESS KEY'
          required
          disabled
          readOnly
          placeholder='MOCK PLACEHOLDER'
          onChange={jest.fn()}
        />
      )

      let instance

      let getIdSpy

      beforeEach(() => {
        jest.clearAllMocks()

        instance = (
          renderer.create(component)
            .getInstance()
        )

        getIdSpy = jest.spyOn(Gremlin.prototype, 'getId').mockReturnValue('MOCK ID')

        instance.renderField()
      })

      it('invokes `getId`', () => {
        return expect(getIdSpy)
          .toBeCalled()
      })

      it('renders `<Field />`', () => {
        return expect(Field)
          .toBeCalledWith({
            name: 'MOCK NAME',
            id: 'MOCK ID',
            value: 'MOCK VALUE',
            tabIndex: 1,
            accessKey: 'MOCK ACCESS KEY',
            required: true,
            disabled: true,
            readOnly: true,
            placeholder: 'MOCK PLACEHOLDER',
            onChange: expect.any(Function)
          }, {})
      })
    })
  })
})
