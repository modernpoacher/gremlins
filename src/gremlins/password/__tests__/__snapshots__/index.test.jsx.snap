// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#gremlins/gremlins/password <Gremlin /> With additional props renders 1`] = `
<div
  className="MOCK CLASSNAME"
/>
`;

exports[`#gremlins/gremlins/password <Gremlin /> With required props renders 1`] = `
<div
  className="MOCK CLASSNAME"
/>
`;
